Implement `is_causal` API for `Transformer`

janEbert opened the pull request:
Adjust documentation according to #97214 and implement `is_causal` API for `Transformer`.
As discussed in #97166.

janEbert committed:
Adjust `*is_causal` documentation

According to #97214, as discussed in #97166.

janEbert committed:
Implement `is_causal` API for `Transformer`

As discussed in #97166.

mikekgfb commented:
Nice! 

Should we add detection of causal mask similar to https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/transformer.py#L302 in TransformerEncoder 
to TransformerDecoder and Transformer as well?

Also, should we replace https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/transformer.py#L305
with a call to Transformer. generate_square_subsequent_mask
(and maybe give that a dtype argument that defaults to torch.float, too?

mikekgfb commented on torch/nn/modules/transformer.py:
```diff
@@ -85,7 +85,8 @@ def __init__(self, d_model: int = 512, nhead: int = 8, num_encoder_layers: int =
                 memory_mask: Optional[Tensor] = None, src_key_padding_mask: Optional[Tensor] = None,
-                tgt_key_padding_mask: Optional[Tensor] = None, memory_key_padding_mask: Optional[Tensor] = None) -> Tensor:
+                tgt_key_padding_mask: Optional[Tensor] = None, memory_key_padding_mask: Optional[Tensor] = None,
+                src_is_causal: bool = False, tgt_is_causal: bool = False, memory_is_causal: bool = False) -> Tensor:
```
Should this be Optional[bool] = None ?
and add detection of causal mask similar to https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/transformer.py#L302 in TransformerEncoder
to TransformerDecoder and Transformer as well?
(We can't do thi at the layer level because it's prohibitively expensive, but we can "buy back" the expense of detecting a causal mask if we get a substantial use out of it.  Users can prevent running this code by setting to True/False.



Also, should we replace https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/transformer.py#L305
with a call to Transformer. generate_square_subsequent_mask
(and maybe give that a dtype argument that defaults to torch.float, too?

janEbert replied:
Agree with everything you said; the un-avoidable causal mask detection and not using `genarate_square_subsequent_mask` had bothered me from the PyTorch 2.0 release as well. :)
Also the missing `dtype` argument had always been a thorn.

That's why I love the recent change of not _always_ trying to detect a causal mask by making `is_causal` `Optional`. I hadn't noticed this change while making the PR, but will definitely integrate your suggestions. Thanks!

Is it fine if I just put everything into this PR or would you like me to make different ones (e.g. for the `generate_square_subsequent_mask` `dtype` argument)?

mikekgfb replied:
I’m good with just amending this the current PR to include these functions

mikekgfb commented:
@janEbert to fix the build fails, please apply this diff - should remove the text commented out, which I did to highlight the change:
```
[mikekg@devgpu001.cln5 ~/fbsource/fbcode/caffe2/test]$ cat /tmp/x
diff --git a/fbcode/caffe2/test/functorch/test_aotdispatch.py b/fbcode/caffe2/test/functorch/test_aotdispatch.py
--- a/fbcode/caffe2/test/functorch/test_aotdispatch.py
+++ b/fbcode/caffe2/test/functorch/test_aotdispatch.py
@@ -2823,14 +2823,14 @@
     torch.nn.TransformerEncoder,  # DataDependentOutputException: aten.equal compares a mask input
                                   # to a causal mask tensor, to see if Boolean is_causal should be set
                                   # for TrnasformerEncoder layers, MHA and sdp custom kernels
-    torch.nn.Transformer,  # DataDependentOutputException: aten.equal compares a mask input
+#    torch.nn.Transformer,  # DataDependentOutputException: aten.equal compares a mask input
                            # to a causal mask tensor, to see if Boolean is_causal should be set
                            # for TransformerEncoder layers, MHA and sdp custom kernels
                            # (this bubbles up to Transformer)
 })

 symbolic_aot_autograd_module_failures = {
-    torch.nn.Transformer,  # DataDependentOutputException: aten.equal compares a mask input to a mask producing a bool
+#    torch.nn.Transformer,  # DataDependentOutputException: aten.equal compares a mask input to a mask producing a bool
     torch.nn.TransformerEncoder,  # DataDependentOutputException: aten.equal compares a mask input to a mask producing a bool
     torch.nn.GaussianNLLLoss,  # NotImplementedError: local_scalar_dense/item NYI for torch.bool
     torch.nn.CrossEntropyLoss,  # Cannot call sizes() on tensor with symbolic sizes/strides
```

janEbert committed:
Adjust `generate_square_subsequent_mask` default device

Now uses `None`, which is more in line with the rest of the codebase and
respects the currently chosen default device.

janEbert committed:
Add dtype argument for `generate_square_subsequent_mask`

janEbert committed:
Fix up docstrings

Some arguments were in the wrong order and whitespace was missing.

janEbert committed:
Fix `TransformerEncoder` `is_causal` documentation

The default was not specific enough and now we explain what it is doing.

janEbert committed:
Refactor causal mask detection

Now uses `generate_square_subsequent_mask`.

janEbert committed:
Merge `if`-statements

No need for the double indent.

janEbert committed:
Remove redundant `if`-statement

`make_causal` is always `False` in this `if`-block already (due to (a)
the prior comparison `make_causal = is_causal is True` and (b) the query
`if is_causal is None`), so we cannot set it to the wrong value.

janEbert committed:
Add causal mask detection for `TransformerDecoder`

janEbert committed:
Refactor causal mask detection

janEbert committed:
Add causal mask detection for `Transformer`

janEbert commented:
Rebased on `main` and applied your suggestions. I took the liberty to include some other minor changes like minor refactoring and documentation adjustments (some arguments were in the wrong order).

I don't manage to build the current head for testing; do I still need to apply your patch even though I implemented the causal mask detection?

janEbert committed:
Fix handling of batched masks

For forward-compatilibity, we also take into account masks being 4-D.

janEbert committed:
Add type annotations for TorchScript

janEbert committed:
Fix failure cause

No longer `aten::equal`; now `aten::eq`.

janEbert committed:
Add `TransformerDecoder` to failures

janEbert committed:
Make type checker happy

janEbert committed:
Add `ModuleInfo` for `TransformerDecoder`

janEbert reviewed (commented):
~~e3afc1a694055046e7b4ea9db1922f50a5d97c5b is the first one that breaks, it seems. I have absolutely no idea what could be the problem.~~

janEbert commented:
Actually nevermind, it seems you somehow ended up on a wrong link to the commits or something. I have no idea.
Anyway, this link with the same commit hash works: https://github.com/pytorch/pytorch/pull/98327/commits/ed37254f86558585f3df2dd8a217435dfbf72272

mikekgfb commented:
Looks great.  The fails seem to be here, in line 3792 of test_nn.
You can presumably make it pass by switching raises=False. 
But wiould also be interesting to know why did it change?

```
        # Incorrect tgt_mask
        encoder_input_shape = correct_encoder_input_shape
        decoder_input_shape = correct_decoder_input_shape
        wrong_tgt_mask_size = tgt_len + 1
        test(encoder_input_shape, decoder_input_shape, tgt_mask_len=wrong_tgt_mask_size, raises=True)
```

janEbert commented:
> One possibility might be that if we are passing a causal mask in, then the flag is_causal=True can preempt the mask from being used, and hence not be diagnosed of being of the wrong size? (The description is very clear -- the is_causal flag can preempt inspection of the attention mask. We lose the performance benefit that the is_causal boolean promises if we're bound to using the attention mask regardless! So, if the is_causal preempts the recognizing of the mask being wrong then in that case, this is the "expected" behavior, in that is_causal can preempt inspection/use of the mask itself.)

Ahh really great point, that seems very likely. So should I rather hardcode `is_causal=False` for these test instead of hoping that the current behavior of the mask shape being ignored stays?

mikekgfb commented:
Yeah, let’s try that. If not, I’m not above just fixing up the test to not raise an exception —- although that no-exception behavior really depends on the hardware this runs, since it requires xformers or flash with only runs on more recent hardware.


Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: janEbert ***@***.***>
Sent: Wednesday, April 19, 2023 12:29:34 AM
To: pytorch/pytorch ***@***.***>
Cc: Michael Gschwind ***@***.***>; Mention ***@***.***>
Subject: Re: [pytorch/pytorch] Implement `is_causal` API for `Transformer` (PR #98327)

One possibility might be that if we are passing a causal mask in, then the flag is_causal=True can preempt the mask from being used, and hence not be diagnosed of being of the wrong size? (The description is very clear -- the is_causal flag
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender

ZjQcmQRYFpfptBannerEnd

One possibility might be that if we are passing a causal mask in, then the flag is_causal=True can preempt the mask from being used, and hence not be diagnosed of being of the wrong size? (The description is very clear -- the is_causal flag can preempt inspection of the attention mask. We lose the performance benefit that the is_causal boolean promises if we're bound to using the attention mask regardless! So, if the is_causal preempts the recognizing of the mask being wrong then in that case, this is the "expected" behavior, in that is_causal can preempt inspection/use of the mask itself.)

Ahh really great point, that seems very likely. So should I rather hardcode is_causal=False for these test instead of hoping that the current behavior of the mask shape being ignored stays?

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/98327#issuecomment-1514260719>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOT4XHPZLE6BISZZW6VRWSDXB6H55ANCNFSM6AAAAAAWTACSVM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

janEbert committed:
Force consideration of attn masks for arg tests

janEbert committed:
Assume error for `src_mask` arg test

This didn't fail (probably indeterministically) due to the causal
attention mask detection.

janEbert commented:
I think this is good now, PTAL. :)

mikekgfb commented:
> I think this is good now, PTAL. :)

Thanks.  This look good.  I've started to split the PR into several pieces, to make review and landing easier as we have a lot of changes going on in this part of PyTorch.  The doc updates are about to land with https://github.com/pytorch/pytorch/pull/101089 and we'll have additional pieces landing in the future!

Thanks for putting these changes together, they make the code much more readable, consistent and better structured!

janEbert commented:
@mikekgfb should I rebase this onto `main` to remove the refactoring part of this? Or do you already have other pieces of this ready so a rebase from my side would be redundant?
